### PR TITLE
Backport: nested function name canonicalization

### PR DESCRIPTION
**Pull request by kpamnany**
## PR Description

Backports https://github.com/JuliaLang/julia/pull/53719.

## Checklist

Requirements for merging:
- [X] I have opened an issue or PR upstream on JuliaLang/julia: https://github.com/JuliaLang/julia/pull/53719
- [X] I have removed the `port-to-*` labels that don't apply.
- [X] I have opened a PR on raicode to test these changes: https://github.com/RelationalAI/raicode/pull/20955

### CODE REVIEW

File: src/julia-syntax.scm [hunk 1]
```diff
@@ -4034,22 +4053,22 @@ f(x) = yt(x)
...
                                         ,body)))))
                        (else
-                        (let* ((exprs     (lift-toplevel (convert-lambda lam2 '|#anon| #t '() #f)))
+                        (let* ((exprs     (lift-toplevel (convert-lambda lam2 '|#anon| #t '() #f parsed-method-stack)))
                                (top-stmts (cdr exprs))
                                (newlam    (compact-and-renumber (linearize (car exprs)) 'none 0)))
                           `(toplevel-butfirst
                             (block ,@sp-inits
-                                   (method ,name ,(cl-convert sig fname lam namemap defined toplevel interp opaq globals locals)
+                                   (method ,(cadr e) ,(cl-convert sig fname lam namemap defined toplevel interp opaq parsed-method-stack globals locals)
```
Comment (d-netto) [inline src/julia-syntax.scm:4071]: Perhaps replace `(cadr e)` with `name` to keep it closer to the original?

Looking at the [v1.10 code](https://github.com/JuliaLang/julia/blob/6f3fdf7b36250fb95f512a2b927ad2518c07d2b5/src/julia-syntax.scm#L4052), seems like the patch which replaced `name` with `(cadr e)` was introduced later.

You will probably need to check whether this works first.

**Review by d-netto — COMMENTED** [review]
Have a minor comment.

I'd be more confident on this backport if you could run the JuliaLang tests locally on your machine and report the results here.

**Comment by kpamnany** [comment]
Some failures:
```
<snip>
asyncmap                                         (15) |    17.35 |   0.38 |  2.2 |     682.96 |   579.92                                                                               
SparseArrays/linalg_solvers                      (15) |        started at 2024-09-12T14:43:22.234                                                                                      
show                                             (31) |         failed at 2024-09-12T14:43:22.912                                                                                      
Test Failed at /home/ubuntu/julia.rai/test/show.jl:790                                                                                                                                 
  Expression: occursin(r"var\"#foo2##[0-9]+", sprint(show, (M53719.foo2(1))(2), context = :module => M53719))                                                                          
   Evaluated: occursin(r"var\"#foo2##[0-9]+", "var\"#bar2##1#bar2##2\"{Int64, Int64}(2, 1)")                                                                                           
                                                                                                                                                                                       
Test Failed at /home/ubuntu/julia.rai/test/show.jl:801                                                                                                                                 
  Expression: occursin(r"var\"#bar#foo##[0-9]+", sprint(show, typeof(M53719.foo(1)), context = :module => M53719))                                                                     
   Evaluated: occursin(r"var\"#bar#foo##[0-9]+", "var\"#bar#bar##0\"{Int64}")                                                                                                          
                                                                                                                                                                                       
Test Failed at /home/ubuntu/julia.rai/test/show.jl:803                                                                                                                                 
  Expression: occursin(r"var\"#baz#foo##[0-9]+", sprint(show, typeof((M53719.foo(1))(2)), context = :module => M53719))                                                                
   Evaluated: occursin(r"var\"#baz#foo##[0-9]+", "var\"#baz#bar##1\"{Int64, Int64}")                                                                                                   
                                                                                                                                                                                       
Test Failed at /home/ubuntu/julia.rai/test/show.jl:805                                                                                                                                 
  Expression: occursin(r"var\"#bar2#foo2##[0-9]+", sprint(show, typeof(M53719.foo2(1)), context = :module => M53719))                                                                  
   Evaluated: occursin(r"var\"#bar2#foo2##[0-9]+", "var\"#bar2#bar2##0\"{Int64}")                                                                                                      
                                                                                                                                                                                       
Test Failed at /home/ubuntu/julia.rai/test/show.jl:807                                                                                                                                 
  Expression: occursin(r"var\"#foo2##[0-9]+#foo2##[0-9]+", sprint(show, typeof((M53719.foo2(1))(2)), context = :module => M53719))                                                     
   Evaluated: occursin(r"var\"#foo2##[0-9]+#foo2##[0-9]+", "var\"#bar2##1#bar2##2\"{Int64, Int64}")
<snip>
```

File: src/julia-syntax.scm (hunk 1 above)
Comment (kpamnany) [reply src/julia-syntax.scm]: Changed it to `name`.

**Review by d-netto — APPROVED** [review]
👍.

Please confirm it passes `raicode` CI before merging.

**Pull request merged**